### ISSUE TITLE
Implement Data/API refactoring ADR (Implementation Plan)

### ISSUE THREAD

**Issue by AlexDBlack** [KonduitAI/konduit-serving#298]
This issue is about how we will proceed with implementing this ADR:
https://github.com/KonduitAI/konduit-serving/blob/master/ADRs/0003-Pipeline_API_Rewrite.md

For the _what_ of the implementation - see the ADR. This issue focuses exclusively on the _how_ of actually executing the implementation.

---
---

Implementing the Pipeline API Rewrite ADR will take a significant amount of time and will impact virtually all aspects of Konduit Serving.

During the buildout/implementation, we want to avoid the following:
* Breaking Konduit Serving on an implementation branch and then not having anything work (or be mergeable) until everything is entirely complete
* Having a huge PR that can't be reviewed effectively due to its large size
* Merge conflicts if master branch is continued to be updated in parallel
* Difficulties in parallelizing the build-out between different engineers due to a monolithic, single-branch rewrite approach.

**The plan**: we will have the new API and old API functionality co-exist in parallel in the master branch of the repository during the duration of ADR implementation.
The modifications for the new API will be worked on without breaking the existing functionality/tests, with code duplication in additional modules/packages when necessary. The new functionality merged frequently, in stages, even if it's not "ready" - other than for new tests, the new API functionality will not be used by Konduit Serving users until the refactoring is complete.
Once the new API functionality is complete and is adequately tested, we can delete the old classes and make it "live".

The main downside of this approach is code duplication.

The actual build-out will proceed in phases, and can be performed in parallel:
Step 1 -> 3 -> (4, 5, 6) -> (7,8)
Step 2 can be done at any point

**Step 1**: Add Data interface, JData and Protobuf implementations
Currently underway.

~~**Step 2**: Deprecate API classes/interfaces, etc to be removed in the new buildout~~
~~outputAdapter, MultiOutputAdapter, Batchoutput, Input, Output, SchemaType, BatchOutput, TensorDataType~~
Complete: https://github.com/KonduitAI/konduit-serving/pull/299

~~**Step 3**: create new pipeline API module; duplicate and update the API for main interfaces
Specifically, duplicate the core interfaces in konduit-serving-api~~
Complete: https://github.com/KonduitAI/konduit-serving/pull/299

We will add a new module: konduit-serving-pipeline
In it, put a new copy of the following API classes/interfaces in ai.konduit.serving.api: PipelineStep, PipelineStepRunner, and Pipeline
Note we will keep out of konduit-serving-pipeline module free of all dependencies such as deeplearning4j, vertx, jpmml, etc.

No changes or copying should be needed for anything that is marked as deprecated in step 2.
Also no changes needed to the following: MetricsConfig, Textconfig and maybe a few others

Why split this up into konduit-serving-pipeline? I want to separate out pipeline API and execution from serving.
This will give us additional flexibility. For example, by avoiding dependencies such as vertex we can easily embed pipeline execution in other non-REST deployment scenarios (in streaming applications like Kafka, spark streaming, etc)

Given the refactoring we're doing (and the staged approach proposed here) we might as well do this part properly at the same time, rather than "building it wrong and fixing it later".

**Step 4**: Update the pipeline steps to use the new API, with one or more steps (/models) in each PR
- ModelStep
    - ~~DL4J~~
    - SameDiff
    - Keras
    - TensorFlow
    - ONNX
    - PMML
- PythonStep
- TransformProcessStep
- ImageLoadingStep
- ArrayConcatenationStep

Note: for ModelStep, we'll flatten the individual classes (flatten InferenceExecutioner, ExecutionerFactory, etc) to 2 levels, from the current 4 in the process
	

**Step 5**: Refactor output adapters to become new pipeline steps

~~**Step 6**: Write new PipelineExecutioner~~
https://github.com/KonduitAI/konduit-serving/tree/master/konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/pipeline

**Step 7**: Write new InferenceVerticle
TBD, but we may do this in a new module: konduit-serving-vertx

**Step 8**: Update CLI

**Step 9**: Delete old functionality and "go live"
